### PR TITLE
[NFR] Added Phalcon\Acl\Adapter\Memory::addRole support multiple inherited

### PR DESCRIPTION
**Pull request by cq-z**
Hello!

* Type: new feature
* Link to issue:

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)?
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.

Small description of change:

Added Phalcon\Acl\Adapter\Memory::addRole support multiple inherited

Thanks

### CODE REVIEW

**Comment by niden** [comment]
@cq-z A couple of tests are failing for PHP 5.x.

Also can you offer a couple of lines on what this does so that we can update the documentation?

Thanks!

**Comment by cq-z** [comment]
@niden does ？I don't quite understand！

**Comment by cq-z** [comment]
@niden There is no 5.x failure  local test environment 。Haven't found the BUG。T-T

**Review by sergeyklay — DISMISSED** [review]
LGTM

**Comment by sergeyklay** [comment]
@cq-z Due to the fact that the 4.x branch is developed separately, could you send send the same PR to the 4.0.x branch ?

**Comment by cq-z** [comment]
sorry the Pull Requests error.    ci is fail.

**Comment by niden** [comment]
@cq-z I think it was a pull issue between Travis / Github. If you want just restart that build that would be great. The message I see is this

```
 cd phalcon/cphalcon
4.70s$ git fetch origin +refs/pull/13556/merge:
fatal: Couldn't find remote ref refs/pull/13556/merge
```

Maybe make a small change, say add a comment and push again to this PR? That will trigger the build fresh.

**Comment by cq-z** [comment]
@niden What is error with this?

```
error: RPC failed; curl 56 OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 10054
Command exited with code 128
PrintLogs "${Env:APPVEYOR_BUILD_FOLDER}\build\php7\safe"
PrintLogs : The term 'PrintLogs' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ PrintLogs "${Env:APPVEYOR_BUILD_FOLDER}\build\php7\safe"

```

**Comment by niden** [comment]
@cq-z Honestly I don't know. I spoke to @sergeyklay and he told me that he knew about that but did not have time to address it.

We should be good once @Jurigag approves this.

File: phalcon/acl/adapter/memory.zep [hunk 1]
```diff
@@ -780,6 +749,88 @@ class Memory extends Adapter
...
+        if isset accessList[accessKey] {
+            return accessKey;
+        }
+        /**
+         * Deep check if the role to inherit is valid
+         */
+        if isset this->_roleInherits[roleName] {
+            let checkRoleToInherits = [];
+            for usedRoleToInherit in this->_roleInherits[roleName] {
+                array_push(checkRoleToInherits,usedRoleToInherit);
```
Comment (Jurigag) [inline phalcon/acl/adapter/memory.zep]: Use `let checkRoleToInherits[] =`

File: phalcon/acl/adapter/memory.zep [hunk 2]
```diff
@@ -780,6 +749,88 @@ class Memory extends Adapter
...
+            }
+            let usedRoleToInherits = [];
+            while !empty checkRoleToInherits {
+                let checkRoleToInherit = array_shift(checkRoleToInherits);
+
+                if isset usedRoleToInherits[checkRoleToInherit] {
+                    continue;
+                }
+                let usedRoleToInherits[checkRoleToInherit]=true;
```
Comment (Jurigag) [inline phalcon/acl/adapter/memory.zep]: spacing

File: phalcon/acl/adapter/memory.zep [hunk 3]
```diff
@@ -780,6 +749,88 @@ class Memory extends Adapter
...
+                let accessKey = checkRoleToInherit . "!*!*";
+                if isset accessList[accessKey] {
+                    return accessKey;
+                }
+                /**
+                 * Push inherited roles
+                 */
+                if isset this->_roleInherits[checkRoleToInherit] {
+                    for usedRoleToInherit in this->_roleInherits[checkRoleToInherit] {
+                        array_push(checkRoleToInherits,usedRoleToInherit);
```
Comment (Jurigag) [inline phalcon/acl/adapter/memory.zep]: spacing, don't use array_push

File: phalcon/acl/adapter/memory.zep [hunk 4]
```diff
@@ -216,49 +217,96 @@ class Memory extends Adapter
...
+                    let usedRoleToInherits[checkRoleToInherit]=true;
+                    if roleName == checkRoleToInherit {
+                        throw new Exception("Role '" . roleInheritName . "' (to inherit) is infinite loop ");
+                    }
+                    /**
+                     * Push inherited roles
+                     */
+                    if isset this->_roleInherits[checkRoleToInherit] {
+                        for usedRoleToInherit in this->_roleInherits[checkRoleToInherit] {
+                            array_push(checkRoleToInherits,usedRoleToInherit);
```
Comment (Jurigag) [inline phalcon/acl/adapter/memory.zep]: Use `let checkRoleToInherits[]`

File: phalcon/acl/adapter/memory.zep [hunk 5]
```diff
@@ -216,49 +217,96 @@ class Memory extends Adapter
...
+                    array_push(checkRoleToInherits,usedRoleToInherit);
+                }
+                let usedRoleToInherits = [];
+                while !empty checkRoleToInherits {
+                    let checkRoleToInherit = array_shift(checkRoleToInherits);
+                    
+                    if isset usedRoleToInherits[checkRoleToInherit] {
+                        continue;
+                    }
+                    let usedRoleToInherits[checkRoleToInherit]=true;
```
Comment (Jurigag) [inline phalcon/acl/adapter/memory.zep]: spacing

File: phalcon/acl/adapter/memory.zep [hunk 6]
```diff
@@ -216,49 +217,96 @@ class Memory extends Adapter
...
+            if roleName == roleInheritName {
+                return false;
+            }
+            /**
+             * Deep check if the role to inherit is valid
+             */
+            if isset this->_roleInherits[roleInheritName] {
+                let checkRoleToInherits = [];
+                for usedRoleToInherit in this->_roleInherits[roleInheritName] {
+                    array_push(checkRoleToInherits,usedRoleToInherit);
```
Comment (Jurigag) [inline phalcon/acl/adapter/memory.zep]: Use `let checkRoleToInherits[] = `

**Review by Jurigag — CHANGES_REQUESTED** [review]
Check comments

File: phalcon/acl/adapter/memory.zep (hunk 1 above)
Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: let checkRoleToInherits[] = in php5.x is error
when array_​shift checkRoleToInherits the this->_roleInherits[roleName] is delete too。
look commit 168f51c

Comment (Jurigag) [reply phalcon/acl/adapter/memory.zep]: But you don't use there [], you use there +, for sure we add in other places lements to array like `let array[] = `

File: phalcon/acl/adapter/memory.zep (hunk 6 above)
Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: It is ok now?

File: phalcon/acl/adapter/memory.zep (hunk 5 above)
Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: It is ok now?

File: phalcon/acl/adapter/memory.zep (hunk 4 above)
Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: It is ok now?

File: phalcon/acl/adapter/memory.zep (hunk 3 above)
Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: It is ok now?

File: phalcon/acl/adapter/memory.zep (hunk 2 above)
Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: It is ok now?

File: phalcon/acl/adapter/memory.zep (hunk 1 above)
Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: It is ok now?

File: phalcon/acl/adapter/memory.zep [hunk 7]
```diff
@@ -780,6 +749,88 @@ class Memory extends Adapter
...
+         * Deep check if the role to inherit is valid
+         */
+        if isset this->_roleInherits[roleName] {
+            let checkRoleToInherits = [];
+            for usedRoleToInherit in this->_roleInherits[roleName] {
+                let checkRoleToInherits[] = usedRoleToInherit;
+            }
+            let usedRoleToInherits = [];
+            while !empty checkRoleToInherits {
+                let checkRoleToInherit = array_shift(checkRoleToInherits);
```
Comment (Jurigag) [inline phalcon/acl/adapter/memory.zep]: Not sure also about this, isn't checkRoleToInherits[0] and unset if necessary enough?

Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: this is  a stack, need popping。

Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: What's the situation now?

Comment (cq-z) [reply phalcon/acl/adapter/memory.zep]: @Jurigag This creates the risk of an infinite loop！
```
let checkRoleToInheritsIndex = 0;
                while !empty checkRoleToInherits {
                    if empty(checkRoleToInherits[checkRoleToInheritsIndex]) {
                        let checkRoleToInheritsIndex ++;
                        continue;
                    }
                    let checkRoleToInherit = checkRoleToInherits[checkRoleToInheritsIndex];
                    unset checkRoleToInherits[checkRoleToInheritsIndex];
                    let checkRoleToInheritsIndex ++;
```

**Comment by niden** [comment]
We need to redo this for the v4 branch. I will try and merge it next week in v4.

**Comment by cq-z** [comment]
@niden 
https://github.com/phalcon/cphalcon/pull/13557
It has been merged

**Comment by niden** [comment]
@cq-z Thanks. I missed that. 

Closing because it was resolved in https://github.com/phalcon/cphalcon/pull/13557

**Pull request closed without merging**